lib/puppet/pops/types/types.rb

```diff
@@ -1,415 +1,1482 @@
+          assignable?(o.type || PUndefType::DEFAULT)
+        else
+          PVariantType::DATA.assignable?(o)
         end
       end
     end
 
-    class PVariantType < PAnyType
-      module ClassModule
+    # Type that is PDataType compatible, but is not a PCollectionType.
+    # @api public
+    #
+    class PScalarType < PAnyType
 
-        def hash
-          [self.class, Set.new(self.types)].hash
-        end
+      def instance?(o)
+        assignable?(TypeCalculator.infer(o))
+      end
 
-        def ==(o)
-          (self.class == o.class && Set.new(types) == Set.new(o.types)) ||
-            (o.class == PDataType && self == Puppet::Pops::Types::TypeCalculator.data_variant())
-        end
+      DEFAULT = PScalarType.new
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        o.is_a?(PScalarType)
       end
     end
 
+    # A string type describing the set of strings having one of the given values
+    # @api public
+    #
     class PEnumType < PScalarType
-      module ClassModule
-        def hash
-          [self.class, Set.new(self.values)].hash
+      attr_reader :values
+
+      def initialize(values)
+        @values = values.sort.freeze
+      end
+
+      def hash
+        @values.hash
+      end
+
+      def ==(o)
+        self.class == o.class && @values == o.values
+      end
+
+      DEFAULT = PEnumType.new([])
+
+      protected
+
+      # @api private
+      def _assignable?(o)
+        return true if self == o
+        svalues = values
+        if svalues.empty?
+          return true if o.is_a?(PStringType) || o.is_a?(PEnumType) || o.is_a?(PPatternType)
+        end
+        case o
+          when PStringType
+            # if the set of strings are all found in the set of enums
+            !o.values.empty? && o.values.all? { |s| svalues.any? { |e| e == s }}
+          when PEnumType
+            !o.values.empty? && o.values.all? { |s| svalues.any? {|e| e == s }}
+          else
+            false
         end
+      end
+    end
 
-        def ==(o)
-          self.class == o.class && Set.new(values) == Set.new(o.values)
+    # @api public
+    #
+    class PNumericType < PScalarType
+      def initialize(from, to = Float::INFINITY)
+        from = -Float::INFINITY if from.nil? || from == :default
+        to = Float::INFINITY if to.nil? || to == :default
+
+        # Always create in right order
+        if from <= to
+          @from = from
+          @to = to
+        else
+          @to = from
+          @from = to
         end
       end
+
+      def from
+        @from == -Float::INFINITY ? nil : @from
+      end
+
+      def to
+        @to == Float::INFINITY ? nil : @to
+      end
+
+      def hash
+        @from.hash * 31 + @to.hash
+      end
+
+      def ==(o)
+        self.class == o.class && @from == o.numeric_from && @to == o.numeric_to
+      end
+
+      def instance?(o)
+        o.is_a?(Numeric) && o >= @from && o <= @to
+      end
+
+      def unbounded?
```

Comment:
IIRC, enumerable? is the only use case for unbounded? in the first place. It would also be neat to be able to iterate over type parameters - e.g. Tuple[Integer, String] would yield Integer, String, and index would yield 0, Integer, 1, String. An Integer[1,2] would yield something like 'min', 1 and 'max',2. Basically representing themselves as either Array or Hash and then iterating over them. -- All of that is worth a separate ticket.

Regarding enumerable? (I am not sure if that is part of a Ruby protocol, and I am not sure we want to be able to iterate over everything that Ruby says is enumerable? true. 

Seems reasonable to keep unbounded? to check lower and upper bounds at the same time, but I like the idea of enumerable?